viewer/packages/cad-geometry-loaders/src/sector/culling/WeightFunctionsHelper.ts

```diff
@@ -0,0 +1,162 @@
+      const frustum = new THREE.Frustum().setFromProjectionMatrix(frustumMatrix);
+      return {
+        ...x,
+        frustum
+      };
+    });
+  }
+
+  addCandidateSectors(sectors: SectorMetadata[], modelMatrix: THREE.Matrix4) {
+    // Note! We compute distance to camera, not screen (which would probably better)
+    const { minDistance, maxDistance } = sectors.reduce(
+      (minMax, sector) => {
+        const distanceToCamera = this.distanceToCamera(sector, modelMatrix);
+        minMax.maxDistance = Math.max(minMax.maxDistance, distanceToCamera);
+        minMax.minDistance = Math.min(minMax.minDistance, distanceToCamera);
+        return minMax;
+      },
+      { minDistance: Infinity, maxDistance: -Infinity }
+    );
+    this._minSectorDistance = minDistance;
+    this._maxSectorDistance = maxDistance;
+  }
+
+  computeTransformedSectorBounds(sectorBounds: THREE.Box3, modelMatrix: THREE.Matrix4, out: THREE.Box3): void {
+    out.copy(sectorBounds);
+    out.applyMatrix4(modelMatrix);
+  }
+
+  /**
+   * Computes a weight in range [0-1], where 1 means close to camera and 0 means far away.
+   */
+  computeDistanceToCameraWeight(transformedSectorBounds: THREE.Box3): number {
+    const minSectorDistance = this._minSectorDistance;
+    const maxSectorDistance = this._maxSectorDistance;
+
+    // Weight sectors that are close to the camera higher
+    const distanceToCamera = transformedSectorBounds.distanceToPoint(this._camera.position);
+    const normalizedDistanceToCamera = (distanceToCamera - minSectorDistance) / (maxSectorDistance - minSectorDistance);
+    return 1.0 - normalizedDistanceToCamera;
+  }
+
+  /**
+   * Compute a weight in range [0-1] where 1 means the sector covers the entire screen
+   * and 0 means no coverage.
+   * @param transformedSectorBounds
+   * @returns
+   */
+  computeScreenAreaWeight(transformedSectorBounds: THREE.Box3): number {
+    const distanceToCamera = transformedSectorBounds.distanceToPoint(this._camera.position);
+    return distanceToCamera > 0.0 ? computeNdcAreaOfBox(this._camera, transformedSectorBounds) : 1.0;
+  }
+
+  /**
+   * Compute a weight in range [0-1] based on at what "depths"
+   * in the frustum the sector is placed.
+   * @param transformedSectorBounds
+   */
+  computeFrustumDepthWeight(transformedSectorBounds: THREE.Box3): number {
+    const frustumWeight = this._modifiedFrustums.reduce((accumulatedWeight, x) => {
+      const { frustum, weight } = x;
+      const accepted = frustum.intersectsBox(transformedSectorBounds);
+      return accumulatedWeight + (accepted ? weight : 0);
+    }, 0.0);
+    return frustumWeight;
+  }
+
+  /**
+   * Computes a weight based on placement in sector tree (i.e. prioritize)
+   * sectors right below root sector.
+   * @param sector
+   */
+  computeSectorTreePlacementWeight(sector: SectorMetadata): number {
+    return sector.depth === 2 ? 1.0 : 1.0 / 3.0;
```

Comment:
This seems kinda brittle and is VERY tied to the tree. I assume any hierarchical change here would totally brake this. If the hierarchy is set in stone then ok, but at least add a comment here (why exactly does depth 2 get a higher weight?).

Reply:
Added comment